### ISSUE TITLE
How to deal with commercial licences

### ISSUE THREAD

**Issue by Bertg** [codepunkt/webpack-license-plugin#1061]
Let me preface this issue whit this. I've read the REASME, and understand that this tool focusses specifically on OSS licenses. However, I think there should be a clean, package supported way to deal with commercial licences. 

We've recently been tasked with creating an inventory of the OSS software packaged in our website. Thanks to this plugin that is a very easy task. It was a bit annoying to have to write the `excludedPackageTest` function to ignore the few commercial packages we have, but nothing to bad. Until I realised I had removed a hole bunch of OSS packages as well.

In this situation we use several packages created by https://fontawesome.com, their OSS packages are MIT (or variations) but we also use some of their commercial packages. These packages all have a very similar name: `@fortawesome/fontawesome-svg-core` for an OSS and `@fortawesome/pro-regular-svg-icons` for a commercial one. 

You can already see where I went wrong. Instead of filtering `packageName.startsWith('@fortawesome/pro-')` I wrote `packageName.startsWith('@fortawesome/')` an easy mistake to make.

# Possible solutions

I think there should be a way, similar to `licenseOverrides` to indicate commercially licensed packages. There are a few ways I could think of:

* Allow `COMMERCIAL` (or similar) as a licence type. (NPM allows `UNLICENSED` and is used by font awesome) (Some people, including myself, are not fan of this, as it is very close to the [Unlicense](https://spdx.org/licenses/Unlicense.html) SPDX identifier)
* Making the above rule even more flexible, a user could define a list of other valid licence names, beyond SPDX.
* Going in a stricter direction, we could have a commercial equivalent of `licenseOverrides`: `commercialLicences`. This could be an object with keys similar to `licenseOverrides` and the values being a string (or function that returns a string) that fetches the specific `licenseText`. 
* If we want to simply exclude commercial packages, I think a stricter option of `excludedPackageTest` can also be created. This could be an option `commercialPackages`, accepting an array of full package names. 

If the maintainers can agree on an approach, I'm willing to put the effort to create the PR.

**Comment by codepunkt** [comment]
@Bertg Thanks for opening and the in-depth description of your struggles. Before I saw you closing the ticket again, I wanted to answer something along the lines of "I fail to see how what you propose would be a substantial benefit compared to the configuration options that are already available".

How'd you solve your current problems?

**Comment by Bertg** [comment]
@codepunkt 

Yeah, the reason I closed the ticket is because I didn't fully understand the outcome of the `excludedPackageTest`. 

I now know how to [add commercial licences](https://github.com/codepunkt/webpack-license-plugin/issues/460#issuecomment-814103682). Honestly, that does seem like a lot of work to make sure that the licence attribution is done correctly. With packages like `@fortawesome/pro*` and `mapbox-gl` (commonly used ones I'm sure) it would be really nice if we could just add them to the list of `licenseOverrides` and allow explicitly the value `UNLICENCED`. 

The problem I described where I accidentally filtered out OSS licences because their similar name as their commercial ones, does seem like a design consideration for this package. A much safer alternative would be to have something like `excludedPackages` an array of packages (with or without version) to be ignored, instead of a function.

Anyhow, I also get the opposite argument. By making the developer experience a bit harder, it forces the developer to think on how to treat their commercial licences.

So yeah, the combination of `excludedPackageTest` and adding licences explicitly is what I'm doing now. I think the DX can be better (possibly just a section in the readme on how to use non OSS licences could help as well) but I can live with it.
